### PR TITLE
[WIP] Replace moderngpu sort join with thrust sort join

### PR DESCRIPTION
**Pull request by kaatish**
ModernGPU based sort join is replace with thrust based sort join.

### CODE REVIEW

File: libgdf/src/join/sort/sort-join.cuh [hunk 1]
```diff
@@ -1,246 +1,227 @@
...
+        gdf_column * const lcol,
+        gdf_column * const rcol,
+        bool flip = false) {
+    int column_width_bytes{0};
+    gdf_error gdf_status = get_column_byte_width(lcol, &column_width_bytes);
+
+    if(GDF_SUCCESS != gdf_status)
+        return gdf_status;
+
+    switch (column_width_bytes) {
```
Comment (jrhemstad) [inline libgdf/src/join/sort/sort-join.cuh]: This `switch` is problematic. 

By casting everything to an `int*_t`, this will technically produce incorrect orderings for unsigned ints as well as floating point values when there is a mix of positive/negative values, as well as sentinel values such as `NaN`. However, I believe for the purposes of this algorithm this is technically _fine_ because you're just trying to group together bit-wise equal keys, and by casting them all to `int` will do this. 

However, where this breaks down is when the type of the column isn't a primitive type (e.g., GDF_STRING) in which case an overloaded `operator<` would need to be used.

In short, I would say this `switch` should be updated to dispatch the correct underlying type, and not just the byte-width. The `gdf_type_dispatcher` I've created should be used for this (though it is not yet merged). See here: https://github.com/rapidsai/cudf/blob/fb8163df529574e3c75603a5e3c14d3164dc3245/libgdf/src/gdf_type_dispatcher.cuh

File: libgdf/src/join/full_join.cuh [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+ */
+
+#include <thrust/copy.h>
+#include <thrust/execution_policy.h>
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+
+/* --------------------------------------------------------------------------*/
+/**
+* @Synopsis  Creates a vector of indices that do not appear in index_ptr
```
Comment (jrhemstad) [inline libgdf/src/join/full_join.cuh]: Could you expand the description here to explain what it means for a value to be missing from the input array?

File: libgdf/src/join/full_join.cuh [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        const size_type buffer_size,
+        const size_type expand_size) {
+    size_type requested_size = buffer_size + expand_size;
+    //No need to proceed if the buffer can contain requested additional elements
+    if (*buffer_capacity >= requested_size) {
+        return GDF_SUCCESS;
+    }
+    data_type * new_buffer{nullptr};
+    data_type * old_buffer = *buffer;
+    CUDA_TRY( cudaMalloc(&new_buffer, requested_size*sizeof(data_type)) );
```
Comment (jrhemstad) [inline libgdf/src/join/full_join.cuh]: This is a common functionality that will likely be used outside of the join algorithm, and therefore I think this entire function could be replaced by a generic `resize` function in RMM. It would take an existing allocation, a new size, and whatever else RMM would need to properly resize it. Thoughts @harrism ?

File: libgdf/src/join/full_join.cuh [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <thrust/copy.h>
+#include <thrust/execution_policy.h>
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+
```
Comment (jrhemstad) [inline libgdf/src/join/full_join.cuh]: It would be super helpful to add a file description and give a high-level description of the full-join algorithm you've implemented. See here: https://github.com/rapidsai/cudf/blob/master/libgdf/src/example_documentation.cpp#L25

File: libgdf/src/join/full_join.cuh (hunk 3 above)
Comment (kaatish) [reply libgdf/src/join/full_join.cuh]: It is slightly different than a resize. The buffer is only expanded if the requested size is greater than its current size.

File: libgdf/src/join/sort/sort-join.cuh (hunk 1 above)
Comment (kaatish) [reply libgdf/src/join/sort/sort-join.cuh]: Since the dispatcher has not been merged yet I have reverted to the old style switch case dependent on the column type. Perhaps this function can be changed once it is in master.

File: libgdf/src/join/full_join.cuh (hunk 4 above)
Comment (kaatish) [reply libgdf/src/join/full_join.cuh]: Added

File: libgdf/src/join/full_join.cuh (hunk 3 above)
Comment (jrhemstad) [reply libgdf/src/join/full_join.cuh]: I believe RMM will track the size of the allocation? And can implement the same logic in which it only expands the allocation if necessary.

File: libgdf/src/join/sort/sort-join.cuh [hunk 5]
```diff
@@ -1,246 +1,230 @@
...
-           comp_t comp, context_t& context)
-{
-    _join_bounds bounds = compute_join_bounds(a, a_count, b, b_count, comp, context);
-    int join_count;
-    mem_t<int> scanned_sizes = scan_join_bounds(bounds, a_count, b_count, context, true,
-                                                join_count);
-    auto output = compute_joined_indices(bounds, scanned_sizes, a_count,
-                                               join_count, context, true);
-    return output;
+compute_joined_indices(const JoinBounds<index_type>& bounds,
```
Comment (jrhemstad) [inline libgdf/src/join/sort/sort-join.cuh]: Missing documentation.

File: libgdf/src/join/sort/sort-join.cuh [hunk 6]
```diff
@@ -1,246 +1,230 @@
...
-    transform_scan<int>([=]MGPU_DEVICE(int index){
-        return upper_rev_data[index] == lower_rev_data[index];
-    }, b_count, matches.data(), plus_t<int>(), matches_count.data(), context);
-
-    // Prepare output
-    append_count = from_mem(matches_count)[0];
-    return matches;
+template <typename index_type>
+void
+create_load_balanced_tuple(const thrust::device_vector<index_type>& scanned_sizes,
```
Comment (jrhemstad) [inline libgdf/src/join/sort/sort-join.cuh]: Missing documentation.

File: libgdf/src/join/sort/sort-join.cuh [hunk 7]
```diff
@@ -1,246 +1,230 @@
...
-    auto appender = [=]MGPU_DEVICE(int index, int seg, int rank) {
-        output_l_data[index + join_count] = -1;
-        output_r_data[index + join_count] = seg;
-    };
-    transform_lbs<launch_arg_t>(appender, append_count, matches.data(),
-                                matches.size(), context);
+template<JoinType join_type,
+    typename index_type>
+thrust::device_vector<index_type>
+scan_join_bounds(const JoinBounds<index_type>& bounds) {
```
Comment (jrhemstad) [inline libgdf/src/join/sort/sort-join.cuh]: Missing documentation.

File: libgdf/src/join/sort/sort-join.cuh [hunk 8]
```diff
@@ -1,246 +1,230 @@
...
-            auto upper = get<1>(lower_upper);
-            auto result = lower + rank;
-            if ( lower == upper ) result = -1;
-            output_l_ptr[index] = seg;
-            output_r_ptr[index] = result;
-        };
-        transform_lbs<launch_arg_t>(k, join_count, scanned_sizes.data(), a_count,
-                                    make_tuple(lower_data, upper_data), context);
+template <typename index_type>
+struct JoinConditionalAdd {
```
Comment (jrhemstad) [inline libgdf/src/join/sort/sort-join.cuh]: Documentation.

File: libgdf/src/join/sort/sort-join.cuh [hunk 9]
```diff
@@ -1,246 +1,230 @@
...
+        return (thrust::get<0>(t) - thrust::get<1>(t)) + (thrust::get<0>(t) == thrust::get<1>(t));
     }
+};
 
-    // Prepare output
-    out_join_count = from_mem(count)[0];
-    return scanned_sizes;
-}
+template <typename index_type>
+struct Diff<index_type, JoinType::INNER_JOIN> {
```
Comment (jrhemstad) [inline libgdf/src/join/sort/sort-join.cuh]: Documentation.

File: libgdf/src/join/sort/sort-join.cuh [hunk 10]
```diff
@@ -1,246 +1,230 @@
...
-    _join_bounds bounds;
-    lower.swap(bounds.lower);
-    upper.swap(bounds.upper);
+//Compute the lower and upper bound
+//Looking for l in r
+template<typename T,
+    typename index_type,
+    typename size_type>
+JoinBounds<index_type>
+compute_join_bounds(
```
Comment (jrhemstad) [inline libgdf/src/join/sort/sort-join.cuh]: Documentation.

File: libgdf/src/join/sort/sort-join.cuh [hunk 11]
```diff
@@ -1,246 +1,230 @@
...
-    int join_count;
-    mem_t<int> scanned_sizes = scan_join_bounds(bounds, a_count, b_count, context, false,
-                                                join_count);
-    auto output = compute_joined_indices(bounds, scanned_sizes, a_count,
-                                               join_count, context, false, 0);
-    return output;
+template<JoinType join_type,
+    typename column_type,
+    typename index_type>
+gdf_error sort_join_typed(
```
Comment (jrhemstad) [inline libgdf/src/join/sort/sort-join.cuh]: Documentation

File: libgdf/src/join/sort/sort-join.cuh [hunk 12]
```diff
@@ -1,246 +1,230 @@
...
-    auto output = compute_joined_indices(bounds, scanned_sizes, a_count,
-                                               join_count, context, false, append_count);
-    full_join_append_right(
-            static_cast<int*>(output.first.data),
-            static_cast<int*>(output.second.data),
-            matches, append_count, join_count, context);
-    return output;
+template<JoinType join_type,
+    typename index_type>
+gdf_error compute_sort_join(
```
Comment (jrhemstad) [inline libgdf/src/join/sort/sort-join.cuh]: Documentation.

File: libgdf/src/join/join_types.h [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JOIN_TYPES
+#define JOIN_TYPES
+
+constexpr int JoinNoneValue = -1;
+
+enum class JoinType {
```
Comment (jrhemstad) [inline libgdf/src/join/join_types.h]: Since you moved these, could you also add documentation?

File: libgdf/src/join/join_types.h [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JOIN_TYPES
+#define JOIN_TYPES
+
+constexpr int JoinNoneValue = -1;
```
Comment (jrhemstad) [inline libgdf/src/join/join_types.h]: Since this was moved, could you also document?

File: libgdf/src/join/full_join.cuh (hunk 3 above)
Comment (harrism) [reply libgdf/src/join/full_join.cuh]: rmmRealloc, although it is currently broken (doesn't do the copy!)

File: libgdf/src/join/sort/sort-join.cuh (hunk 1 above)
Comment (harrism) [reply libgdf/src/join/sort/sort-join.cuh]: The dispatcher is merged.

**Comment by kaatish** [comment]
rerun tests

**Comment by kaatish** [comment]
rerun tests

**Comment by kaatish** [comment]
run tests

File: cpp/src/join/full_join.cuh [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+            *l_index_ptr + *index_size,
+            *l_index_ptr + *index_size + mismatch_index_size,
+            JoinNoneValue);
+
+    //Copy unmatched indices to the r_index_ptr
+    thrust::copy(thrust::device,
+            unmatched_indices.begin(),
+            unmatched_indices.begin() + mismatch_index_size,
+            *r_index_ptr + *index_size);
+    *index_capacity = l_index_capacity;
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]: Are you writing to a `const` parameter here?

File: cpp/src/join/full_join.cuh [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+* @Returns  cudaSuccess upon successful completion of append call. Otherwise returns
+* the appropriate CUDA error code
+*/
+/* ----------------------------------------------------------------------------*/
+template <typename index_type, typename size_type>
+gdf_error append_full_join_indices(
+        index_type ** l_index_ptr,
+        index_type ** r_index_ptr,
+        size_type * const index_capacity,
+        size_type * const index_size,
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]: If this is a pointer to a scalar value, shouldn't you just pass a reference instead?

File: cpp/src/join/full_join.cuh [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+    //Copy unmatched indices to the r_index_ptr
+    thrust::copy(thrust::device,
+            unmatched_indices.begin(),
+            unmatched_indices.begin() + mismatch_index_size,
+            *r_index_ptr + *index_size);
+    *index_capacity = l_index_capacity;
+    *index_size = *index_size + mismatch_index_size;
+
+    CUDA_CHECK_LAST()
+	return GDF_SUCCESS;
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]:
```suggestion
    return GDF_SUCCESS;
```

File: cpp/src/join/full_join.cuh [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+
+    //Copy JoinNoneValue to l_index_ptr to denote that a match does not exist on the left
+    thrust::fill(
+            thrust::device,
+            *l_index_ptr + *index_size,
+            *l_index_ptr + *index_size + mismatch_index_size,
+            JoinNoneValue);
+
+    //Copy unmatched indices to the r_index_ptr
+    thrust::copy(thrust::device,
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]: Be consistent with argument alignment. Notice difference from this line and line 165.

File: cpp/src/join/full_join.cuh [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+    size_type l_index_capacity = *index_capacity;
+    size_type r_index_capacity = *index_capacity;
+    err = expand_buffer(l_index_ptr, &l_index_capacity, *index_size, mismatch_index_size);
+    if (GDF_SUCCESS != err) return err;
+    err = expand_buffer(r_index_ptr, &r_index_capacity, *index_size, mismatch_index_size);
+    if (GDF_SUCCESS != err) return err;
+
+    //Copy JoinNoneValue to l_index_ptr to denote that a match does not exist on the left
+    thrust::fill(
+            thrust::device,
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]: See above comments about using an RMM allocator.

File: cpp/src/join/joining.h [hunk 20]
```diff
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+#pragma once
```
Comment (harrism) [inline cpp/src/join/joining.h]: Please use #ifdef/#ifndef. #pragma once is not part of the C++ standard.

File: cpp/src/join/sort_join.cuh [hunk 21]
```diff
@@ -1,247 +1,325 @@
...
+    typename size_type>
+JoinBounds<index_type>
+compute_join_bounds(
+        T const * const l, size_type l_count,
+        T const * const r, size_type r_count) {
+    JoinBounds<index_type> bounds;
+    bounds.lower.resize(l_count);
+    bounds.upper.resize(l_count);
+    thrust::lower_bound(
+            thrust::device,
```
Comment (harrism) [inline cpp/src/join/sort_join.cuh]: Please use an RMM allocator for these Thrust algorithms to optimize intermediate allocations.

File: cpp/src/join/sort_join.cuh [hunk 22]
```diff
@@ -1,247 +1,325 @@
...
+ * @tparam index_type Output type for the index calculation
+ *
+ */
+/* ----------------------------------------------------------------------------*/
+template <typename index_type>
+void
+create_load_balanced_tuple(const thrust::device_vector<index_type>& scanned_sizes,
+        index_type * const seg, index_type * const rank, const index_type segment_length) {
+    thrust::upper_bound(
+            thrust::device,
```
Comment (harrism) [inline cpp/src/join/sort_join.cuh]: Please use an RMM allocator for these thrust algorithms.

File: cpp/src/join/sort_join.cuh [hunk 23]
```diff
@@ -1,247 +1,325 @@
...
+ * @Param seg Array containing repeated indices
+ * @Param rank Array containing repeation rank of seg
+ * @Param segment_length Last value in scanned_sizes
+ * @tparam index_type Output type for the index calculation
+ *
+ */
+/* ----------------------------------------------------------------------------*/
+template <typename index_type>
+void
+create_load_balanced_tuple(const thrust::device_vector<index_type>& scanned_sizes,
```
Comment (harrism) [inline cpp/src/join/sort_join.cuh]: Please use RMM allocator for device_vector.

File: cpp/src/join/sort_join.cuh [hunk 24]
```diff
@@ -1,247 +1,325 @@
...
-    _join_bounds bounds = compute_join_bounds(a, a_count, b, b_count, comp, context);
-    int join_count;
-    mem_t<int> scanned_sizes = scan_join_bounds(bounds, a_count, b_count, context, false,
-                                                join_count);
-    auto output = compute_joined_indices(bounds, scanned_sizes, a_count,
-                                               join_count, context, false, 0);
-    return output;
+    index_type * l_ptr;
+    index_type * r_ptr;
+    cudaMalloc(&l_ptr, join_size*sizeof(index_type));
```
Comment (harrism) [inline cpp/src/join/sort_join.cuh]: Use RMM. Don't call cudaMalloc or cudaFree directly in cuDF.

File: cpp/src/join/sort_join.cuh [hunk 25]
```diff
@@ -1,247 +1,325 @@
...
+        gdf_column * const output_l,
+        gdf_column * const output_r,
+        gdf_column * const leftcol,
+        gdf_column * const rightcol,
+        bool flip_results = false) {
+    JoinBounds<index_type> bounds =
+        compute_join_bounds<column_type, index_type>(
+                static_cast<column_type*>(leftcol->data), leftcol->size,
+                static_cast<column_type*>(rightcol->data), rightcol->size);
+    thrust::device_vector<index_type> scanned_sizes =
```
Comment (harrism) [inline cpp/src/join/sort_join.cuh]: Use RMM allocator.

File: cpp/tests/join/join_tests.cu [hunk 26]
```diff
@@ -580,9 +580,9 @@ typedef ::testing::Types<
                           TestParameters< join_op::FULL, HASH, VTuple<int32_t > >,
                           TestParameters< join_op::FULL, HASH, VTuple<int64_t > >,
                           TestParameters< join_op::FULL, HASH, VTuple<float   > >,
-                          TestParameters< join_op::FULL, HASH, VTuple<double  > >,
```
Comment (harrism) [inline cpp/tests/join/join_tests.cu]: Are you removing hash-based FULL join tests for these three datatypes? Shouldn't you just be adding tests for sort-based join, not removing the hash-based tests?

File: cpp/src/join/full_join.cuh [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+* @Param max_index_value The maximum value an index can have in index_ptr
+* @Param index_size Number of left and right indices
+* @tparam index_type The type of data associated with index_ptr
+* @tparam size_type The data type used for size calculations
+*
+* @Returns  thrust::device_vector containing the indices that are missing from index_ptr
+*/
+/* ----------------------------------------------------------------------------*/
+template <typename index_type, typename size_type>
+thrust::device_vector<index_type>
```
[the remark this answers is not in the document]
Comment (kaatish) [reply cpp/src/join/full_join.cuh]: RVO will take care of this.

File: cpp/src/join/full_join.cuh [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+        const size_type buffer_size,
+        const size_type expand_size) {
+    size_type requested_size = buffer_size + expand_size;
+    //No need to proceed if the buffer can contain requested additional elements
+    if (*buffer_capacity >= requested_size) {
+        return GDF_SUCCESS;
+    }
+    data_type * new_buffer{nullptr};
+    data_type * old_buffer = *buffer;
+    CUDA_TRY( cudaMalloc(&new_buffer, requested_size*sizeof(data_type)) );
```
[the remark this answers is not in the document]
Comment (kaatish) [reply cpp/src/join/full_join.cuh]: I have replaced cudaMallocs and cudaFrees with corresponding RMM calls for now

File: cpp/src/join/full_join.cuh [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+	size_type compacted_size = thrust::copy_if(
+			thrust::device,
+			thrust::make_counting_iterator(begin_counter),
+			thrust::make_counting_iterator(end_counter),
+			invalid_index_map.begin(),
+			unmatched_indices.begin(),
+			thrust::identity<index_type>()) -
+		unmatched_indices.begin();
+	unmatched_indices.resize(compacted_size);
+	return unmatched_indices;
```
[the remark this answers is not in the document]
Comment (kaatish) [reply cpp/src/join/full_join.cuh]: There is a CUDA_CHECK_LAST() call after this function is called in append_full_join_indices.

File: cpp/src/join/full_join.cuh [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+template <typename index_type, typename size_type>
+gdf_error append_full_join_indices(
+        index_type ** l_index_ptr,
+        index_type ** r_index_ptr,
+        size_type * const index_capacity,
+        size_type * const index_size,
+        const size_type max_index_value) {
+    gdf_error err;
+    //Get array of indices that do not appear in r_index_ptr
+    thrust::device_vector<index_type> unmatched_indices =
```
[the remark this answers is not in the document]
Comment (kaatish) [reply cpp/src/join/full_join.cuh]: Assignment operator will not be called here. RVO will take care of this.

File: cpp/src/join/full_join.cuh (hunk 15 above)
Comment (kaatish) [reply cpp/src/join/full_join.cuh]: The pointer is const. The value being pointed by it isn't.

File: cpp/tests/join/join_tests.cu (hunk 26 above)
Comment (kaatish) [reply cpp/tests/join/join_tests.cu]: Unfortunately, there is a limit to the number of template arguments in testing::Types. A few lines were removed to avoid compilation errors.

File: cpp/src/join/full_join.cuh [hunk 31]
```diff
@@ -0,0 +1,194 @@
...
+/** ---------------------------------------------------------------------------*
+* @file full_join.cuh
+* @brief Implementation of full join.
+*
+* The functions in this file are used to append indices to the output of a left
+* join call to create the output of a full join. The highest level function in
+* this file is append_full_join_indices.
+* ---------------------------------------------------------------------------**/
+
+template <typename T>
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]: You can delete these two lines and just use rmm::device_vector now.

File: cpp/src/join/full_join.cuh [hunk 32]
```diff
@@ -0,0 +1,194 @@
...
+/* ----------------------------------------------------------------------------*/
+template <typename index_type, typename size_type, typename ExecutionPolicy>
+DeviceVector<index_type>
+create_missing_indices(
+        index_type const * const index_ptr,
+        const size_type max_index_value,
+        const size_type index_size,
+        const ExecutionPolicy& exec_policy) {
+    //Assume all the indices in invalid_index_map are invalid
+    DeviceVector<index_type> invalid_index_map(max_index_value, 1);
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]:
```suggestion
    rmm::device_vector<index_type> invalid_index_map(max_index_value, 1);
```

File: cpp/src/join/full_join.cuh [hunk 33]
```diff
@@ -0,0 +1,194 @@
...
+    //Vector allocated for unmatched result
+    DeviceVector<index_type> unmatched_indices(max_index_value);
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]:
```suggestion
   rmm::device_vector<index_type> unmatched_indices(max_index_value);
```

File: cpp/src/join/full_join.cuh [hunk 34]
```diff
@@ -0,0 +1,194 @@
...
+        index_type ** l_index_ptr,
+        index_type ** r_index_ptr,
+        size_type &index_capacity,
+        size_type &index_size,
+        const size_type max_index_value,
+        MemAlloc& allocator,
+        cudaStream_t stream) {
+    gdf_error err;
+    //Get array of indices that do not appear in r_index_ptr
+    DeviceVector<index_type> unmatched_indices =
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]:
```suggestion
    rmm::device_vector<index_type> unmatched_indices =
```

File: cpp/src/join/join_compute_api.h [hunk 35]
```diff
@@ -340,7 +195,8 @@ gdf_error estimate_join_output_size(gdf_table<size_type> const & build_table,
 /* ----------------------------------------------------------------------------*/
 template<JoinType join_type,
          typename output_index_type,
-         typename size_type>
+         typename size_type,
+         typename MemAlloc = rmm_temp_allocator>
```
Comment (harrism) [inline cpp/src/join/join_compute_api.h]: This template parameter is unused? Any reason to template the allocator?

File: cpp/src/join/full_join.cuh [hunk 36]
```diff
@@ -0,0 +1,194 @@
...
+    size_type l_index_capacity = index_capacity;
+    size_type r_index_capacity = index_capacity;
+    err = expand_buffer(l_index_ptr, &l_index_capacity, index_size, mismatch_index_size, stream);
+    if (GDF_SUCCESS != err) return err;
+    err = expand_buffer(r_index_ptr, &r_index_capacity, index_size, mismatch_index_size, stream);
+    if (GDF_SUCCESS != err) return err;
+
+    //Copy JoinNoneValue to l_index_ptr to denote that a match does not exist on the left
+    thrust::fill(
+            thrust::cuda::par(allocator).on(stream),
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]: Please use `rmm::exec_policy(stream)`. See https://github.com/rapidsai/cudf/blob/3350027fff3960dac269eba652bd217e497a9640/cpp/src/rmm/thrust_rmm_allocator.h#L93

```suggestion
            rmm::exec_policy(stream),
```

Then you can get rid of the allocator parameter to this function (there was no reason to have this parameter anyway, since instantiating the allocator has no cost).

File: cpp/src/join/full_join.cuh [hunk 37]
```diff
@@ -0,0 +1,194 @@
...
+            *l_index_ptr + index_size,
+            *l_index_ptr + index_size + mismatch_index_size,
+            JoinNoneValue);
+
+    //Copy unmatched indices to the r_index_ptr
+    thrust::copy(
+            thrust::cuda::par(allocator).on(stream),
```
Comment (harrism) [inline cpp/src/join/full_join.cuh]:
```suggestion
            rmm::exec_policy(stream),
```

File: cpp/src/join/join_compute_api.h (hunk 35 above)
Comment (harrism) [reply cpp/src/join/join_compute_api.h]: Pretty sure you can remove this.

File: cpp/src/join/join_compute_api.h [hunk 38]
```diff
@@ -507,11 +363,14 @@ gdf_error compute_hash_join(
   // free memory used for the counters
   RMM_TRY( RMM_FREE(d_global_write_index, 0) );
 
+  cudaStream_t stream = 0;
+  rmm_temp_allocator allocator(stream);
```
Comment (harrism) [inline cpp/src/join/join_compute_api.h]: You can drop this, and the allocator parameter to `append_full_join_indices`.

File: cpp/tests/join/join_tests.cu (hunk 26 above)
Comment (harrism) [reply cpp/tests/join/join_tests.cu]: Sounds reasonable. I don't want to leave functionality untested.

File: cpp/src/join/sort_join.cuh [hunk 39]
```diff
@@ -1,247 +1,345 @@
...
-> (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
-> SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
->
-> The views and conclusions contained in the software and documentation are those
-> of the authors and should not be interpreted as representing official policies,
-> either expressed or implied, of the FreeBSD Project.
-*/
-
-/* Sort-based join using moderngpu */
+ * Copyright (c) 2017, NVIDIA CORPORATION.
```
Comment (harrism) [inline cpp/src/join/sort_join.cuh]: Why did you add a 2017 copyright? Please update
.

File: cpp/src/join/sort_join.cuh [hunk 40]
```diff
@@ -1,247 +1,345 @@
...
+    typename size_type>
+JoinBounds<index_type>
+compute_join_bounds(
+        T const * const l, size_type l_count,
+        T const * const r, size_type r_count,
+        cudaStream_t stream) {
+    JoinBounds<index_type> bounds;
+    bounds.lower.resize(l_count);
+    bounds.upper.resize(l_count);
+    thrust::lower_bound(
```
Comment (harrism) [inline cpp/src/join/sort_join.cuh]: On my machine this doesn't compile because you didn't `#include <thrust/binary_search.h>`. Please add that include.

Comment (kaatish) [reply cpp/src/join/sort_join.cuh]: Added

**Comment by dantegd** [comment]
@harrism @kaatish @jrhemstad  I haven't gone through the C++ side of the code yet, but a quick debugging from python/cython reveals the following:

Using this portion of the code in one of the pytests:

```
aa = [0, 0, 4, 5, 5]
bb = [0, 0, 2, 3, 5]

df = cudf.DataFrame()
df['a'] = aa
df['b'] = bb

df1 = df.set_index('a')
df2 = df.set_index('b')
joined = df1.join(df2, how='left', sort=True, method='hash')
```

We would expect the following result: 

```
   b    a
0  0  0.0
0  0  0.0
0  0  0.0
0  0  0.0
4  2  NaN
5  3  5.0
5  5  5.0
```

So after checking that th values get sent to C++ correctly (which they do as far as I can see), instead we get the following *data* raw results from the `libcudf.gdf_left_join`: 

```
b: [0 0 3 5 2 0 0]
a: [0 0 5 5 4 0 0]
index: [                  0                   0                   5
                   5 9187201950435737471                   0
                   0]
```

Note that the index is just the column `on` which the merge happens. So it seems that there is a wrong index or small error somewhere when gathering or assigning the values of the `on` column. This could be overflowing or corrupting the memory of the valid masks, since trying to print them immediately after the call gives: 

```
CudaAPIError: [1] Call to cuMemcpyDtoH results in CUDA_ERROR_INVALID_VALUE
```

Of course the valid mask error could be unrelated, but it could also be explained by the behavior in the `on` column.

**Comment by jrhemstad** [comment]
rerun tests

**Comment by harrism** [comment]
@kaatish I take it this PR can be closed since you opened PR #1349 ?
